packages/linter/src/utils/rules-requiring-type-checking.ts

```diff
@@ -0,0 +1,89 @@
+import type { Linter } from 'eslint';
+
+// Cache the resolved rules from node_modules
+let knownRulesRequiringTypeChecking: string[] | null = null;
+
+function resolveKnownRulesRequiringTypeChecking(): string[] | null {
+  if (knownRulesRequiringTypeChecking) {
+    return knownRulesRequiringTypeChecking;
+  }
+  try {
+    const { rules } = require('@typescript-eslint/eslint-plugin');
+    const rulesRequiringTypeInfo = Object.entries(rules)
+      .map(([ruleName, config]) => {
+        if ((config as any).meta?.docs?.requiresTypeChecking) {
+          return `@typescript-eslint/${ruleName}`;
+        }
+        return null;
+      })
+      .filter(Boolean);
+    return rulesRequiringTypeInfo;
+  } catch (err) {
+    console.log(err);
+    return null;
+  }
+}
+
+export function hasRulesRequiringTypeChecking(
+  eslintConfig: Linter.Config
+): boolean {
+  knownRulesRequiringTypeChecking = resolveKnownRulesRequiringTypeChecking();
+  if (!knownRulesRequiringTypeChecking) {
+    /**
+     * If (unexpectedly) known rules requiring type checking could not be resolved,
+     * default to assuming that the rules are in use to align most closely with Nx
+     * ESLint configs to date.
+     */
+    return true;
+  }
+  const allRulesInConfig = getAllRulesInConfig(eslintConfig);
+  return allRulesInConfig.some((rule) =>
+    knownRulesRequiringTypeChecking.includes(rule)
+  );
+}
+
+export function removeParserOptionsProjectIfNotRequired(
+  json: Linter.Config
+): Linter.Config {
+  // At least one rule requiring type-checking is in use, do not migrate the config
+  if (hasRulesRequiringTypeChecking(json)) {
+    return json;
+  }
+  removeProjectParserOptionFromConfig(json);
+  return json;
+}
+
+function getAllRulesInConfig(json: Linter.Config): string[] {
+  let allRules = [];
+  if (json.rules) {
+    allRules = [...allRules, ...Object.keys(json.rules)];
+  }
+  if (json.overrides?.length > 0) {
+    for (const override of json.overrides) {
+      if (override.rules) {
+        allRules = [...allRules, ...Object.keys(override.rules)];
+      }
+    }
+  }
+  return allRules;
+}
+
+function removeProjectParserOptionFromConfig(json: Linter.Config): void {
```

Comment:
Is there a reason why this function is not immutable?
Also lines 72-76 and 79-86 are duplicate. We can perhaps write something like:

```typescript
const cleanupProject = (json) => {
  const { parserOptions, ...rest };
  // no parserOptions or project, return the original object
  if (!parserOptions || !parserOptions.project) {
  	return json;
  }
  const { project, ...cleanParserOptions } = parserOptions;
  // if there are some options left
  if (Object.keys(cleanParserOptions).length) {
    return { ...rest, parserOptions: cleanParserOptions };
  }
  return rest;
}
```
And we could then apply it to entire json as well as the overrides (using `map` instead of `for...in`).

Reply:
This very subjective again I’m afraid. Not every line of duplication requires a function and not every instance of mutability is inherently “bad” as you are implying here.

I would say that the code is clear, readable and easy to maintain in its current state - which are the most important factors IMO.

I happy to hear from others on this though

Reply:
That was for...of typo 😀